### PR TITLE
Suspend solution crawler during line commit

### PR DESCRIPTION
**Pull request by rchande**
Line commit on enter goes from ~40 seconds in a 3000 line file to instantaneous.

**Customer scenario**

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

**Bugs this fixes:**

(either VSO or GitHub links)

**Workarounds, if any**

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

**Risk**

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

**Performance impact**

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

**Is this a regression from a previous update?**

**Root cause analysis:**

How did we miss it?  What tests are we adding to guard against it in the future?

**How was the bug found?**

(E.g. customer reported it vs. ad hoc testing)

### CODE REVIEW

**Comment by rchande** [comment]
Tag @dotnet/roslyn-ide for review
This addresses a DTS bug around line commit.

File: src/EditorFeatures/VisualBasic/LineCommit/CommitBufferManager.vb [hunk 1]
```diff
@@ -109,14 +114,18 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.LineCommit
                         Return
                     End If
 
-                    Dim dirtyRegion = _dirtyState.DirtyRegion.GetSpan(_buffer.CurrentSnapshot)
-                    Dim info As FormattingInfo
-                    If Not TryComputeExpandedSpanToFormat(dirtyRegion, info, cancellationToken) Then
-                        Return
-                    End If
+                    Using _notificationService.Start("LineCommit")
+
```
Comment (jasonmalinowski) [inline src/EditorFeatures/VisualBasic/LineCommit/CommitBufferManager.vb]: Blank lines.

File: src/EditorFeatures/VisualBasic/LineCommit/CommitBufferManager.vb [hunk 2]
```diff
@@ -109,14 +114,18 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.LineCommit
                         Return
                     End If
 
-                    Dim dirtyRegion = _dirtyState.DirtyRegion.GetSpan(_buffer.CurrentSnapshot)
-                    Dim info As FormattingInfo
-                    If Not TryComputeExpandedSpanToFormat(dirtyRegion, info, cancellationToken) Then
-                        Return
-                    End If
+                    Using _notificationService.Start("LineCommit")
```
Comment (Pilchie) [inline src/EditorFeatures/VisualBasic/LineCommit/CommitBufferManager.vb]: Does this need to be localized?

**Comment by heejaechang** [comment]
interesting.. disabling solution crawler improves that much with multi core machine? it might be due to task starvation since formatter will try to run concurrently if span it is formatting is big otherwise do it synchronously.

if there are a lot of pending tasks then formatting might get blocked until there is space in task scheduler..

File: src/EditorFeatures/VisualBasic/LineCommit/CommitBufferManager.vb (hunk 2 above)
Comment (rchande) [inline src/EditorFeatures/VisualBasic/LineCommit/CommitBufferManager.vb]: I don't think so.

**Comment by rchande** [comment]
@heejaechang In traces of commit with solution crawler, I see us recovering syntax trees over and over when line commit tries to bind tokens. I'm going to try to determine what's being recovered.

**Comment by CyrusNajmabadi** [comment]
> In traces of commit with solution crawler, I see us recovering syntax trees over and over when line commit tries to bind tokens.

Could a more tactical fix be for us to just get the syntax-root up front, and then hold onto it for the lifetime of the commit operation?

**Comment by rchande** [comment]
@CyrusNajmabadi It looks like all of the code cleanup providers start out by first getting root for the tree they're going to fix up.

**Comment by heejaechang** [comment]
we need better cache. @sharwell I nominate Sam for the task! :)

**Comment by CyrusNajmabadi** [comment]
> It looks like all of the code cleanup providers start out by first getting root for the tree they're going to fix up.

yes.  but we could simply fix that by just having the system get the root first, then call into the providers.  This is what we do in other areas...

**Comment by heejaechang** [comment]
each code cleanup provider transform tree and give it to next one. so I dont think root engine has will matter to later code clean up providers since it is different tree.

**Comment by CyrusNajmabadi** [comment]
@heejaechang I don't understand this bit then:

> I see us recovering syntax trees over and over when line commit tries to bind tokens

When we're binding the tokens, why do we need to recover syntax trees?

**Comment by heejaechang** [comment]
I didn't investigate the issue, so I don't know either. @rchande ?

**Comment by CyrusNajmabadi** [comment]
if that was hte issue, it seems like that would be trivially solvable instead of needing to suspend solution crawler.

**Comment by heejaechang** [comment]
there is 2 way to fork solution, which reference equality or value equality. if solution is forked with value equality, it will persist tree rather than keeping them in memory. @rchande which forking does code clean up do? or is that recovering not about on the file code clean up is running?

**Comment by rchande** [comment]
@heejaechang It calls GetSyntaxRootAsync--which fork type is that? (http://index/?query=casecorrectasync&rightProject=Microsoft.CodeAnalysis.Workspaces&file=CaseCorrection%5CAbstractCaseCorrectionService.cs&line=27)

**Comment by CyrusNajmabadi** [comment]
I'm confused.  Why is that being called multiple times @rchande?  It looks like it gets the root once and then case corrects all the tokens in teh spans specified.  It seems like that would only be getting the root once.

**Pull request merged**